src/protocols/bdx/BdxEndpoint.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "BdxEndpoint.h"
+
+#include <core/CHIPError.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeDelegate.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <protocols/bdx/BdxTransferSession.h>
+#include <support/BitFlags.h>
+#include <system/SystemClock.h>
+#include <system/SystemLayer.h>
+
+namespace chip {
+namespace bdx {
+
+CHIP_ERROR Endpoint::OnMessageReceived(chip::Messaging::ExchangeContext * ec, const chip::PacketHeader & packetHeader,
+                                       const chip::PayloadHeader & payloadHeader, chip::System::PacketBufferHandle && payload)
+{
+    if (mExchangeCtx == nullptr)
+    {
+        mExchangeCtx = ec;
+    }
+
+    ChipLogDetail(BDX, "%s: message 0x%x protocol %u", __FUNCTION__, static_cast<uint8_t>(payloadHeader.GetMessageType()),
+                  payloadHeader.GetProtocolID().GetProtocolId());
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::Platform::Clock::GetMonotonicMilliseconds());
```

Comment:
Do we send a response to this message anywhere?  If so, we need to let the exchange know we plan to do that...  (That said, I don't see where in this PR we do it).

Reply:
Can you tell me more about how that works? I wrote it this way so that a sleepy device could handle this message and sleep for a bit before responding. But all BDX messages will need a response eventually until the end of the transfer.

Reply:
Please see documentation for ExchangeDelegate::OnMessageReceived.  After calling this method, the exchange will close itself unless one of two things has happened:

1. A message was sent and a response is expected.
2. `WillSendMessage` was called on the exchange to indicate that it should stay open pending the message send.

In this case probably we want to call `WillSendMessage` and ensure that consumers know to close this exchange?

Reply:
It looks like my demo branch is somewhat out of date which may be why I didn't notice this.

We should probably let the application logic decide and call `WillSendMessage`, or have an [in/out] parameter `willSendMessage` that the application logic can use to signal back to `Endpoint`.

We could try to deduce `willSendMessage` from from the message type that was received in this message, but I think there might not be enough context to do so accurately. For example, a `StandaloneAck` message could come in the middle of a transfer (where it would expect a response) or at the end (where it wouldn't expect a response).

Reply:
The other option is to just always call `WillSendMessage` here and then do manual `Close` calls as needed if we decide there is no response after all.  Whatever works best; I am not 100% clear on the control flow through the bdx bits yet.

Reply:
I think this makes the most sense. Since most BDX messages are going to have a response, except for a few messages that might signify the end of a transfer.